receiver/datadogreceiver/internal/translator/traces_translator.go

```diff
@@ -102,6 +103,33 @@ func ToTraces(payload *pb.TracerPayload, req *http.Request) ptrace.Traces {
 			_ = tagsToSpanLinks(span.GetMeta(), newSpan.Links())
 
 			newSpan.SetTraceID(uInt64ToTraceID(0, span.TraceID))
+
+			// Datadog traces split a 128 bits trace id in two parts: TraceID and Tags._dd_p_tid. This happens if the
+			// instrumented service received a TraceContext from an OTel instrumented service. When it happens, we need
+			// to concatenate the two into newSpan.TraceID.
+			// The following map keeps track of the TraceIDs we process as only the first span has the upper 64 bits from the
+			// 128 bits trace ID.
+			//
+			// Reconstructing the 128 bits TraceID, if available or cached.
+			// Note: This may not be resilient to related spans being flushed separately in datadog's tracing libraries.
+			//       It might also not work if multiple datadog instrumented services are chained.
+			//
+			// This is currently gated by a feature gate. If we don't get a cache here, we don't enable this behavior.
+			if traceIDCache != nil {
+				if val, ok := span.Meta["_dd.p.tid"]; ok {
+					hexTraceID := val + strconv.FormatUint(span.TraceID, 16)
+					traceID, err := oteltrace.TraceIDFromHex(hexTraceID)
```

Comment:
Shall we enrich the error message with the provided `hexTraceID` to troubleshoot the error `hex encoded trace-id must have length equals to 32`?

Reply:
Added